File: backend/backend/src/main/java/com/moovie/backend/domain/repository/MoviesRepository.java
package com.moovie.backend.domain.repository;

import com.moovie.backend.domain.model.*;
import com.moovie.backend.domain.model.MovieListing;
import com.moovie.backend.domain.model.MovieListListing;
import com.moovie.backend.domain.model.MovieReviewListing;

import java.util.List;

public interface MoviesRepository {

    MovieListing getTrendingMovies(int page);

    MovieDetail getMovieDetail(int movieId);

    MovieListing searchMovies(String query, int page);

    MovieListing discoverMovies(int page, String year, String releaseDateGte, String releaseDateLte,
                                String sortBy, String withGenres, String withOriginalLanguage,
                                String withOriginCountry, String voteCountGte);

    List<Genre> getGenres();

    List<Country> getCountries();

    List<Language> getLanguages();

    MovieReviewListing getMovieReviews(int movieId, int page);

    MovieListing getUserFavoriteMovies(String userId, int page);

    MovieListing getUserWatchList(String userId, int page);

    MovieListListing getMovieLists(int page, String userId);

    MovieListListing getUserMovieLists(int page);

    MovieListDetail getMovieListDetail(int listId, int page);

    MovieListListing getFeaturedLists(int page);
}

File: src/main/kotlin/org/example/domain/repository/MoviesRepository.kt
package org.example.domain.repository

import org.example.domain.model.*
import org.example.model.MovieListing
import org.example.model.MovieListListing
import org.example.model.MovieReviewListing

interface MoviesRepository {
    suspend fun getTrendingMovies(page: Int): MovieListing
    suspend fun getMovieDetail(movieId: Int): MovieDetail
    suspend fun searchMovies(query: String, page: Int): MovieListing
    suspend fun discoverMovies(
        page: Int,
        year: Int? = null,
        releaseDateGte: String? = null,
        releaseDateLte: String? = null,
        sortBy: String? = null,
        genres: String? = null,
        language: String? = null,
        country: String? = null,
        voteCountGte: Int? = null
    ): MovieListing
    suspend fun getGenres(): List<Genre>
    suspend fun getCountries(): List<Country>
    suspend fun getLanguages(): List<Language>
    suspend fun getMovieReviews(page: Int, userId: String?, movieId: Int?): MovieReviewListing
    suspend fun getUserFavoriteMovies(userId: String, page: Int): MovieListing
    suspend fun getUserWatchList(userId: String, page: Int): MovieListing
    suspend fun getMovieLists(page: Int, userId: String?): MovieListListing
    suspend fun getUserMovieLists(page: Int): MovieListListing
    suspend fun getMovieListDetail(listId: Int, page: Int): MovieListDetail
    suspend fun getFeaturedLists(page: Int): MovieListListing
}
